File: Pandacap.LowLevel/Data/Credentials/DeviantArtCredentials.fs
namespace Pandacap.Data

open System.ComponentModel.DataAnnotations
open Pandacap.PlatformBadges

/// The active credentials for the DeviantArt API.
type DeviantArtCredentials() =
    [<Key>]
    member val Username = "" with get, set

    member val AccessToken = "" with get, set
    member val RefreshToken = "" with get, set

    interface DeviantArtFs.IDeviantArtAccessToken with
        member this.AccessToken = this.AccessToken

    interface IExternalCredentials with
        member this.Username = this.Username
        member _.Platform = DeviantArt

File: Pandacap.Database/DeviantArtCredentials.cs
using Pandacap.UI.Badges;
using Pandacap.UI.Elements;
using System.ComponentModel.DataAnnotations;

namespace Pandacap.Database
{
    public class DeviantArtCredentials : IExternalCredentials
    {
        [Key]
        public string Username { get; set; } = "";

        public string AccessToken { get; set; } = "";
        public string RefreshToken { get; set; } = "";

        Badge IExternalCredentials.Badge =>
            Badges.DeviantArt.WithText(Username);
    }
}
